bookkeeper-server/src/main/java/org/apache/bookkeeper/replication/ReplicasMigrationWorker.java

```diff
@@ -0,0 +1,216 @@
+    private static final Logger LOG = LoggerFactory.getLogger(ReplicasMigrationWorker.class);
+    private final BookKeeperAdmin admin;
+    private final BiConsumer<Long, Long> onReadEntryFailureCallback;
+    private final MetadataClientDriver metadataClientDriver;
+    private final String rootPath;
+    private final String replicasMigrationPath;
+    private final String lock = "locked";
+    private final LedgerChecker ledgerChecker;
+    private final Thread workerThread;
+    private final long backoffMs = 100;
+    private volatile boolean workerRunning = false;
+    private final String seperator = ",";
+    private final String advertisedAddress;
+
+    @StatsDoc(
+            name = NUM_ENTRIES_UNABLE_TO_READ_FOR_MIGRATION,
+            help = "the number of entries ReplicasMigrationWorker unable to read"
+    )
+    private final Counter numEntriesUnableToReadForMigration;
+
+    public ReplicasMigrationWorker(final ServerConfiguration conf, BookKeeper bkc, StatsLogger statsLogger) {
+        this.advertisedAddress = conf.getAdvertisedAddress();
+        this.rootPath = ZKMetadataDriverBase.resolveZkLedgersRootPath(conf);
+        this.replicasMigrationPath = getReplicasMigrationPath(rootPath);
+        this.ledgerChecker = new LedgerChecker(bkc);
+        this.admin = new BookKeeperAdmin(bkc, statsLogger, new ClientConfiguration(conf));
+        this.metadataClientDriver = bkc.getMetadataClientDriver();
+        this.numEntriesUnableToReadForMigration = statsLogger
+                .getCounter(NUM_ENTRIES_UNABLE_TO_READ_FOR_MIGRATION);
+        this.onReadEntryFailureCallback = (ledgerid, entryid) -> {
+            numEntriesUnableToReadForMigration.inc();
+        };
+        this.workerThread = new BookieThread(this, "ReplicasMigrationWorker");
+    }
+
+    public static String getReplicasMigrationPath(String rootPath) {
+        return String.format("%s/%s", rootPath, BookKeeperConstants.MIGRATION_REPLICAS);
+    }
+
+    public static String getLockForMigrationReplicasPath(String replicasMigrationPath, long ledgerId, String lock) {
+        return String.format("%s/%s/%s", replicasMigrationPath, ledgerId, lock);
+    }
+
+    public static String getLedgerForMigrationReplicasPath(String replicasMigrationPath, long ledgerId) {
+        return String.format("%s/%s", replicasMigrationPath, ledgerId);
+    }
+
+    static class Replicas {
+        long ledgerId;
+        Set<BookieId> bookieIds;
+
+        public Replicas(long ledgerId, Set<BookieId> bookieIds) {
+            this.ledgerId = ledgerId;
+            this.bookieIds = bookieIds;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("<ledgerId=%s,bookies=<%s>", ledgerId, bookieIds);
+        }
+    }
+
+    private Replicas getReplicasToMigrate() throws InterruptedException, KeeperException {
+        List<String> ledgerIds = metadataClientDriver.listLedgersOfMigrationReplicas(replicasMigrationPath);
```

Comment:
We get the whole list for each time and pick one ledger from the earliest, it will cost a lot of resources. Would you consider to tune it?

Reply:
OK， I will fix!

Reply:
Fixed.
PTAL,thanks! @hangc0276